OpenRA.Mods.RA2/Traits/Mirage.cs

```diff
@@ -0,0 +1,250 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.RA2.Traits
+{
+	[Desc("Tag trait for trees used as mirage.")]
+	public class MirageTargetInfo : TraitInfo<MirageTarget> { }
+	public class MirageTarget { }
+
+	[Desc("Overrides the default Tooltip to aid in deceiving enemy players.")]
+	class MirageTooltipInfo : TooltipInfo, Requires<MirageInfo>
+	{
+		public override object Create(ActorInitializer init) { return new MirageTooltip(init.Self, this); }
+	}
+
+	class MirageTooltip : ConditionalTrait<MirageTooltipInfo>, ITooltip
+	{
+		readonly Actor self;
+		readonly Mirage mirage;
+
+		public MirageTooltip(Actor self, MirageTooltipInfo info)
+			: base(info)
+		{
+			this.self = self;
+			mirage = self.Trait<Mirage>();
+		}
+
+		public ITooltipInfo TooltipInfo { get { return Info; } }
+
+		public Player Owner
+		{
+			get
+			{
+				if (!mirage.IsMirage || self.Owner.IsAlliedWith(self.World.RenderPlayer))
+					return self.Owner;
+
+				return self.World.Players.First(p => p.InternalName == mirage.Info.EffectiveOwner);
+			}
+		}
+	}
+
+	[Flags]
+	public enum MirageRevealType
+	{
+		None = 0,
+		Attack = 1,
+		Move = 2,
+		Unload = 4,
+		Infiltrate = 8,
+		Demolish = 16,
+		Damage = 32,
+		Heal = 64,
+		SelfHeal = 128,
+		Dock = 256
+	}
+
+	[Desc("This actor can appear as a different actor in specific situations.")]
+	public class MirageInfo : PausableConditionalTraitInfo
+	{
+		[Desc("Measured in game ticks.")]
+		public readonly int InitialDelay = 10;
+
+		[Desc("Measured in game ticks.")]
+		public readonly int MirageDelay = 30;
```

Comment:
Can we explain what this is in the `Desc`? I suppose this is some kind of `RevealTime`.

Reply:
I renamed it.